server.js

```diff
@@ -2056,14 +2047,57 @@ if (config.startServer) {
           process.exit(0);
         }
 
-        process.on('SIGTERM', () => {
-          opentelemetry
-            .shutdown()
-            .catch((err) => logger.error('Error shutting down OpenTelemetry', err))
-            .finally(() => {
-              process.exit(0);
-            });
-        });
+        const prepareForTermination = async () => {
+          // We want to proceed with termination even if something goes wrong,
+          // so don't allow this function to throw.
+          try {
+            await util.promisify(server.close).call(server);
+
+            await socketServer.close();
+
+            // Wait for all currently-executing cron jobs to finish.
+            await cron.stop();
+
+            // Wait for all currently-executing server jobs to finish.
+            await serverJobs.stop();
+          } catch (err) {
+            logger.error('Error while shutting down server', err);
+            Sentry.captureException(err);
+          }
+        };
+
+        const terminate = async () => {
+          // Shut down OpenTelemetry exporting.
+          try {
+            opentelemetry.shutdown();
+          } catch (err) {
+            logger.error('Error shutting down OpenTelemetry', err);
+            Sentry.captureException(err);
+          }
+
+          // Flush all events to Sentry.
+          try {
+            await Sentry.flush();
+          } finally {
+            process.exit(0);
+          }
+        };
+
+        // If we're running in EC2 auto scaling, this will wait for our instance
+        // to enter the `Terminating:Wait` state.
+        //
+        // If we're not running in EC2, this will be a no-op.
+        lifecycleHooks
+          .handleAndCompleteInstanceTermination(prepareForTermination)
```

Comment:
If I'm understanding correctly, this will poll every 10 seconds forever? Should we instead set up an SNS notification or a lambda that will call a webhook or something like that?

Reply:
I thought about the webhook approach, but I opted to go the simpler route for now.

To think about what that might look like though: when we get notified about an instance being shut down, we'd somehow discover its internal hostname and post directly to port 80 on the host with some kind of secret?

Alternatively... I'm really curious if we could just rely on systemd SIGTERMing us when the instance is being shut down? Presumably shutdown is graceful and not just "whoops we killed power to your box".

Reply:
> I thought about the webhook approach, but I opted to go the simpler route for now.

Understandable. I just can't help but feel like polling is horrible here 😄

> To think about what that might look like though: when we get notified about an instance being shut down, we'd somehow discover its internal hostname and post directly to port 80 on the host with some kind of secret?

Yep, the Event Bridge notification that's sent to the lambda contains the instance id, and from that we can look up the instance public hostname from the instance metadata: https://docs.aws.amazon.com/autoscaling/ec2/userguide/ec2-auto-scaling-event-reference.html

> Alternatively... I'm really curious if we could just rely on systemd SIGTERMing us when the instance is being shut down? Presumably shutdown is graceful and not just "whoops we killed power to your box".

Yep, it's graceful, but [the docs say](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Stop_Start.html#what-happens-stop):
> If the instance OS does not shut down cleanly within a few minutes, a hard shutdown is performed.

I couldn't find anything that specifies what "a few minutes" means.

Reply:
Alright, relying on `SIGTERM` seems like a bad idea.

In the interest of landing something that we can iterate on, are you ok with sticking with polling for now? EventBridge + Lambda seems like an objectively better option, and it'll be important once we get large numbers of chunk servers in operation, but I don't think it's necessary at the moment.

Reply:
No worries, I'm happy to go with whatever you want here. I'm slightly nervous that we'll never implement the EventBridge code if we deploy with polling, because it will be a scary change to launch into prod. But I also agree that setting up EV+Lambda is a pain in the neck (although terraform should make it much better than when I last did this by hand).